### PR TITLE
[Validator] Perform renaming of structs/services and attributes (without test update)

### PR DESCRIPTION
**Pull request by c-harding**
Splitting up #235 by removing all the updates to tests. Everything still passes, it just doesn’t demonstrate any of the pwer of this PR. Check out #235 to see the full gory detail of what happens when you apply it to tests, but basically if any naming conflict is found, it is then resolved by prefixing the name with the project name (repeatedly until no more conflicts).

- `NameClashes`, a new object to handle names reserved by Postgres. This is composable so any lists of names can be used to determine conflicts. **100** lines of this are just a list of Postgres reserved keywords
- `SemanticContext`, removing that this was a function (see `Validator`).
- `ServiceRenamer`, a new class for performing renaming (of services/structs) on a Templefile. It is constructed with a map from old names to new, and then serves as a function from Templefile to Templefile. A lot of this file is exhaustive matches, even when they are just the identity function. This is so that any newly added metadata types etc. will throw MatchErrors here instead of silently allowing unrenamed items through, which would become a real pain to debug.
- `Validator`: The validator now also performs renaming on the Templefile’s fields and attributes (Suggestions of what to rename this class (and its methods) to are welcome!).
  It seemed a bit strange to me that `context` was a function, so now an explicit `foreachValueWithContext` etc is used, serving basically the same purpose but a little more explicitly.

Re codecov, every new untested line is tested in the e2e tests

Spurious branch naming as I cherry-picked my way to two non-XXL PRs.

### CODE REVIEW

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    def fromIterable(strings: Set[String]): NameValidator = NameValidator { name =>
+      !strings.iterator.contains(name.toLowerCase)
+    }
+  }
+
+  val postgresValidator: NameValidator = NameValidator.fromIterable(postgresNames)
+
+  /** Repeatedly add the project name to the start of the string */
+  def nameSuggestions(name: String, project: String, decapitalize: Boolean = false): Iterator[String] =
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala]: This can be private, no?

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    iterate(name.capitalize) { project + _ }.map { if (decapitalize) StringUtils.decapitalize else identity }
+
+  def dodgeNames(name: String, project: String, takenNames: Set[String], decapitalize: Boolean = false)(
+    validators: NameValidator*,
+  ): String = {
+    val validator = NameValidator.combine(Iterator(NameValidator.fromIterable(takenNames)) ++ validators)
+    nameSuggestions(name, project, decapitalize).filter { validator.isValid(_) }.next
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala]: I had to look up how this was implemented to double check it was efficient - from first glance you're filtering over an infinite iterator, which didn't seem right 😆

File: src/main/scala/temple/DSL/semantics/ServiceRenamer.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        case writable: Metadata.Writable        => writable
+      },
+    )
+
+  private def renameTargetBlocks(targets: Map[String, TargetBlock]): Map[String, TargetBlock] =
+    targets.view.mapValues(renameTargetBlock).toMap
+
+  private def renameServiceMetadata(metadata: Metadata.ServiceMetadata): Metadata.ServiceMetadata = metadata match {
+    // rename any services referenced in #uses
+    case Metadata.Uses(services) => Metadata.Uses(services.map(renaming))
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/ServiceRenamer.scala]: This isn't too clear what's going on? You're using a map lookup as a function? Then what happens if `renaming` doesn't contain that key?

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 4]
```diff
@@ -1,44 +1,81 @@
...
 import temple.ast.{Metadata, _}
+import temple.builder.project.ProjectConfig
 
 import scala.collection.mutable
 import scala.reflect.{ClassTag, classTag}
 
-private class Validator private (templefile: Templefile) {
-  var errors: mutable.Set[String] = mutable.SortedSet()
+private class Validator(templefile: Templefile) {
+  var errors: mutable.Buffer[String] = mutable.Buffer()
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala]: Is the mutable prefix needed? Don't think it'd make much sense to have an `immutable.Buffer` 😆

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 5]
```diff
@@ -1,44 +1,81 @@
...
+  private var newServices: Map[String, ServiceBlock] = templefile.services
+
+  private val globalRenaming      = mutable.Map[String, String]()
+  def allGlobalNames: Set[String] = globalRenaming.valuesIterator.toSet ++ allStructsAndServices
+
+  private def validateAttributes(
+    attributes: Map[String, Attribute],
+    context: SemanticContext,
+  ): Map[String, Attribute] = {
+    foreachValueWithContext(attributes, context, validateAttribute)
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala]: There seems to be a lot of words here, but I'm still not quite sure what's going on? Is it really worth the function call or could we just put that single line here?

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 6]
```diff
@@ -1,44 +1,81 @@
...
+    attributes.foldLeft(attributes.mapFactory[String, Attribute]()) {
+      case (newAttributes, (attributeName, value)) =>
+        if (attributeName.headOption.exists(!_.isLower))
+          errors += context.errorMessage(
+            s"Invalid attribute name $attributeName, it must start with a lowercase letter,",
+          )
+        val newName = dodgeNames(
+          attributeName,
+          templefile.projectName,
+          (newAttributes.keys ++ attributes.keys).toSet - attributeName,
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala]: A comment explaining this might be nice

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 7]
```diff
@@ -158,6 +198,38 @@ object Validator {
...
+    validator.transformed
   }
+
+//  private type EntryValidator[T] = (String, T, SemanticContext) => Unit
+//
+//  @inline final private def foreachEntryWithContext[T](
+//    map: Map[String, T],
+//    context: SemanticContext,
+//    validate: EntryValidator[T],
+//  ): Unit = map.foreachEntry { case (name, t) => validate(name, t, context :+ name) }
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala]: Can we clean this up?

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 8]
```diff
@@ -1,44 +1,81 @@
...
+  private val globalRenaming      = mutable.Map[String, String]()
+  def allGlobalNames: Set[String] = globalRenaming.valuesIterator.toSet ++ allStructsAndServices
+
+  private def validateAttributes(
+    attributes: Map[String, Attribute],
+    context: SemanticContext,
+  ): Map[String, Attribute] = {
+    foreachValueWithContext(attributes, context, validateAttribute)
+
+    // This looks like a map, but it isn’t because the accumulator `newAttributes`’ names are accessed during iteration
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala]: Rather than saying what this should or shouldn't look like, maybe explain the map you're building up?

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+
+  val postgresValidator: NameValidator = NameValidator.fromIterable(postgresNames)
+
+  /** Repeatedly add the project name to the start of the string */
+  def nameSuggestions(name: String, project: String, decapitalize: Boolean = false): Iterator[String] =
+    iterate(name.capitalize) { project + _ }.map { if (decapitalize) StringUtils.decapitalize else identity }
+
+  def dodgeNames(name: String, project: String, takenNames: Set[String], decapitalize: Boolean = false)(
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala]: Hmm not sure I like `dodge`

Comment (jaylees14) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: Could this have a comment too? Maybe explaining what happens if the name is or isn't unique? The API usage isn't very clear here

File: src/main/scala/temple/DSL/semantics/NameClashes.scala [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    validators: NameValidator*,
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/NameClashes.scala:141]: Any reason why this was taken separately?

File: src/main/scala/temple/DSL/semantics/Validator.scala [hunk 11]
```diff
@@ -1,44 +1,81 @@
...
+        newAttributes + (newName -> value)
     }
-    attributes.foreachEntry(context(validateAttribute))
   }
 
-  private def validateService(service: ServiceBlock)(context: SemanticContext): Unit = {
-    service.structs.foreachEntry(context(validateStruct))
-    validateAttributes(service.attributes)(context)
-    validateMetadata(service.metadata, service.attributes)(context)
+  private def renameBlock(name: String, block: TempleBlock[_]): Unit = {
```
Comment (jaylees14) [inline src/main/scala/temple/DSL/semantics/Validator.scala:63]: A comment here, explaining the side effect please

**Review by jaylees14 — COMMENTED** [review]
Changes looks mostly understandable. `Validate.scala` has now become pretty complicated, so I think we'd all appreciate a little more documentation in there, especially given that some functions have side effects.

Am I right in saying that there's the assumption that `globalRenaming` will contain exactly the map of everything in the project (that could be renamed) to its new name? If so that part seems pretty hard to follow, and I can't check that that is necessarily true.

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 2 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: What would you suggest I do instead?

File: src/main/scala/temple/DSL/semantics/ServiceRenamer.scala (hunk 3 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/ServiceRenamer.scala]: Yes, maps are functions. I could make it clearer if you’d like. If it doesn’t include the key it will throw a (maybe not particularly easy to read) error, I’ll update that to make it more readable.

File: src/main/scala/temple/DSL/semantics/Validator.scala (hunk 4 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/Validator.scala]: Good point, was a mutable set so I’ll fix that

File: src/main/scala/temple/DSL/semantics/Validator.scala (hunk 5 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/Validator.scala]: My concern was that there would be places where we forgot to append to the context if we did it directly, but I will try to see how it looks

File: src/main/scala/temple/DSL/semantics/Validator.scala (hunk 6 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/Validator.scala]: Good point

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 9 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: Sure, will do. What would you suggest for the name?

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 10 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala:141]: It’s a varargs following an optional parameter, so I’d have to make `decapitalize` required to allow both.

File: src/main/scala/temple/DSL/semantics/Validator.scala (hunk 11 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/Validator.scala:63]: Sure, coming soon

**Comment by c-harding** [comment]
@jaylees14 documentation is on its way. globalRenaming does contain every service and struct, even if they’re not named. I’ll make that more explicit too.

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 2 above)
Comment (jaylees14) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: I think it’s fine as is, just made me think a little 😉

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 10 above)
Comment (jaylees14) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala:141]: Ah fair enough 😀

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 9 above)
Comment (jaylees14) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: Hmm, naming isn’t my strong point. `constructUniqueName`? @smithwjv is usually pretty good at this stuff

Comment (willjvsmith) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: What is it doing

**Review by jaylees14 — APPROVED** [review]
Awesome job @xsanda! Nice and readable, really good stuff

**Review by jaylees14 — APPROVED** [review]

File: src/main/scala/temple/DSL/semantics/NameClashes.scala (hunk 9 above)
Comment (c-harding) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: @smithwjv taking a name for a service/struct/attribute, and changing it if necessary not to conflict with reserved names in any host language

Comment (willjvsmith) [reply src/main/scala/temple/DSL/semantics/NameClashes.scala]: Looks like a `createValidName` or `constructValidName` to me

**Review by Lewiky — APPROVED** [review]
Cool LGTM
